File: openless-all/app/src/lib/ipc/selection-correction.ts
import { invokeOrMock } from './shared';

export type SelectionCorrectionAction = 'literalReplace' | 'review';
export type SelectionCorrectionBubbleState = 'actions' | 'recording' | 'processing' | 'error';

export interface SelectionCorrectionBubblePayload {
  selectedText: string;
  state: SelectionCorrectionBubbleState;
  action: SelectionCorrectionAction | null;
  message: string | null;
}

export function getSelectionCorrection(): Promise<SelectionCorrectionBubblePayload | null> {
  return invokeOrMock('get_selection_correction', undefined, () => ({
    selectedText: 'Codex Context',
    state: 'actions',
    action: null,
    message: null,
  }));
}

export function startSelectionCorrection(action: SelectionCorrectionAction): Promise<void> {
  return invokeOrMock('start_selection_correction', { action }, () => undefined);
}

export function stopSelectionCorrection(): Promise<void> {
  return invokeOrMock('stop_selection_correction', undefined, () => undefined);
}

export function cancelSelectionCorrection(): Promise<void> {
  return invokeOrMock('cancel_selection_correction', undefined, () => undefined);
}

export function dismissSelectionCorrection(): Promise<void> {
  return invokeOrMock('dismiss_selection_correction', undefined, () => undefined);
}

File: openless-all/app/src-tauri/src/commands/selection_correction.rs
use super::*;
use crate::coordinator::selection_correction::{
    SelectionCorrectionAction, SelectionCorrectionBubblePayload,
};

#[tauri::command]
pub fn get_selection_correction(
    coord: CoordinatorState<'_>,
) -> Option<SelectionCorrectionBubblePayload> {
    coord.selection_correction_payload()
}

#[tauri::command]
pub async fn start_selection_correction(
    coord: CoordinatorState<'_>,
    action: SelectionCorrectionAction,
) -> Result<(), String> {
    coord.start_selection_correction(action).await
}

#[tauri::command]
pub async fn stop_selection_correction(coord: CoordinatorState<'_>) -> Result<(), String> {
    coord.stop_selection_correction().await
}

#[tauri::command]
pub fn cancel_selection_correction(coord: CoordinatorState<'_>) {
    coord.cancel_selection_correction();
}

#[tauri::command]
pub fn dismiss_selection_correction(coord: CoordinatorState<'_>) {
    coord.dismiss_selection_correction();
}
